### PR TITLE
[KBFS-1646] Use Block Retrieval Workers in BlockOpsStandard

### PR DESCRIPTION
**Pull request by jzila**

### CODE REVIEW

File: libkbfs/block_retrieval_worker.go [hunk 1]
```diff
@@ -67,8 +67,11 @@ func (brw *blockRetrievalWorker) HandleRequest() (err error) {
 	}
 
 	// Create a new block of the same type as the first request
-	typ := reflect.TypeOf(retrieval.requests[0].block).Elem()
-	block := reflect.New(typ).Interface().(Block)
+	var block Block
+	if retrieval.requests[0].block != nil {
```
Comment (strib) [inline libkbfs/block_retrieval_worker.go]: Do some tests pass in a `nil` block or something?  This seems like an odd check to need.  Seems like the block retriever wouldn't be able to handle a `nil` block anyway.

**Comment by akalin-keybase** [comment]
Before I forget, can you add a few stat counters to see how well the workers are utilized? (In a separate PR.) Probably one gauge for total `BlockOps.Get` requests, and another for requests that are "saved" by grouping with another one. Then we'll have a ratio saying how often the workers save us from multiple fetches.

**Comment by jzila** [comment]
@strib: tests passed, ready for a final look.

File: libkbfs/block_retrieval_worker.go [hunk 2]
```diff
@@ -57,7 +57,12 @@ func (brw *blockRetrievalWorker) HandleRequest() (err error) {
 	}
 
 	// Create a new block of the same type as the first request
-	block := retrieval.requests[0].block.NewEmpty()
+	var block Block
+	retrieval.reqMtx.RLock()
+	if retrieval.requests[0].block != nil {
```
Comment (strib) [inline libkbfs/block_retrieval_worker.go]: What happens when it is `nil`?  I don't think we can safely pass a typeless, `nil` block into `getBlock` below, can we?

File: libkbfs/block_retrieval_worker_test.go [hunk 3]
```diff
@@ -247,4 +248,5 @@ func TestBlockRetrievalWorkerShutdown(t *testing.T) {
 	require.True(t, shutdown)
 	w.Shutdown()
 	require.True(t, shutdown)
+	cancel()
```
Comment (strib) [inline libkbfs/block_retrieval_worker_test.go]: If you keep this, why not a `defer cancel()` right where this is created, so if any early returns are added in between we still cancel it?

File: libkbfs/block_retrieval_worker_test.go [hunk 4]
```diff
@@ -236,7 +236,8 @@ func TestBlockRetrievalWorkerShutdown(t *testing.T) {
 
 	w.Shutdown()
 	block := &FileBlock{}
-	ch := q.Request(context.Background(), 1, nil, ptr1, block)
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (strib) [inline libkbfs/block_retrieval_worker_test.go]: Was this just for testing?  Seems like the `cancel` isn't really used.

Comment (jzila) [inline libkbfs/block_retrieval_worker_test.go]: it's used to ensure that the context goroutine is broken out of its loop, so we don't leak goroutines out of this test.

File: libkbfs/block_retrieval_worker.go (hunk 2 above)
Comment (jzila) [inline libkbfs/block_retrieval_worker.go]: Maybe I should panic here if it's `nil`. That really should never happen.

Comment (jzila) [inline libkbfs/block_retrieval_worker.go]: Hmm. Apparently we have at least one test checking for this. WDYT @strib?
https://ci.keyba.se/job/kbfs/job/PR-480/22/execution/node/238/log/

**Comment by strib** [comment]
On 11/07/2016 06:10 PM, John Zila wrote:

>
> Hmm. Apparently we have at least one test checking for this. WDYT
> @strib <https://github.com/strib>?
> https://ci.keyba.se/job/kbfs/job/PR-480/22/execution/node/238/log/
>
>

Looks like those tests are just being lazy by passing in `nil` blocks. 
Should be trivial to change them to pass in a *TestBlock instead.

**Pull request merged**